### ISSUE TITLE
Twilio Request Validator not Matching Hash

### ISSUE THREAD

**Issue by eaharonoff** [twilio/twilio-php#510]
**Version:**
Composer.json
 "require": {
        "php": ">=7.1",
        "laravel/framework": "5.3.*",
        "aloha/twilio": "^4.0"
    },

### Code Snippet
```php
<?php

namespace App\Http\Middleware;

use Closure;
use App\Http\Responses\Errors;
use Twilio\Security\RequestValidator;
use Log;

class TwilioAuth
{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next)
    {
        try {
            $token = config('constants.twilio.auth_token');
            $signature = $_SERVER["HTTP_X_TWILIO_SIGNATURE"];
            $validator = new RequestValidator($token);
            $isValid = $validator->validate($signature, $request->url(), $request->all());
            if ($isValid) {
                return $next($request);

            }
            else {
                throw new \Exception;
            }
        }
        catch (\Exception $e) {
            $output = [
                'error' => [
                    'message' => "Insufficient privileges.",
                    'code' => Errors::AUTH_TWILIO_ERROR,
                ],
                'data' => $e->getMessage(),
            ];
            return response()->json($output, 403);
        }


    }
}

```

### Exception/Log
```
[2018-10-18 23:55:05] production.WARNING: http://[redacted]/api/v1/twilio/webhooks/opt-out?bodySHA256=f3d4040e206be251a0892b3e2ad43b22a8c08ff3b6793b65eefd648bcee3e5f1(full url)
[2018-10-18 23:55:05] production.WARNING: array (
  'From' => '+1516[redacted]',
  'bodySHA256' => 'f3d4040e206be251a0892b3e2ad43b22a8c08ff3b6793b65eefd648bcee3e5f1',
) (full request body)
[2018-10-18 23:55:05] production.WARNING: d8VYA7BN/sFTTixhfO41jw4udbc= (api side generated key)
[2018-10-18 23:55:05] production.WARNING: TJUXXLZdnLHrFtIBNsQzuhHqntg= (twilio generated header key)
```

I am using **Twilio Studio** widgets in an auto-responder flow and the setup is as follows: 
If I get a certain response (say person X texts "STOP"), I have widget that will trigger a POST request to hit my API endpoint. The problem is I cannot get the twilio generated signature to match. I also find it odd that i have a `bodySHA256` param when I don't see something like that mentioned in the docs. Can someone provide some clarity on what is happening here? Also, let me know if you need me to provide any more info, I've been struggling with this all week 😓 

Attached is the Studio widget tied to this case.
![screen shot 2018-10-18 at 7 53 02 pm](https://user-images.githubusercontent.com/13280719/47190452-81c5e380-d30f-11e8-87e2-ca2e3ff09eda.png)

**Comment by johnpaulmedina** [comment]
I had a similar issue on 5.3 its because of the Request class rearranging the items. Try using $_REQUEST from php.

**Comment by eaharonoff** [comment]
> I had a similar issue on 5.3 its because of the Request class rearranging the items. Try using $_REQUEST from php.

I tried that, and here is what it produced:
```
[2018-10-19 00:07:39] production.WARNING: array (
  'bodySHA256' => 'f3d4040e206be251a0892b3e2ad43b22a8c08ff3b6793b65eefd648bcee3e5f1',
)
```

**Comment by eaharonoff** [comment]
I am also aware there are a few other open issues on this matter (https://github.com/twilio/twilio-php/issues/436) but I have not read anything conclusive on this topic.

**Comment by johnpaulmedina** [comment]
Can you show me the Twilio Console Request and also Log the Request on the Laravel side so that we can compare. Remember the Signature must match exactly. So it will basically take each value in their orders and then hash them - they MUST BE identical on each side. One can not have an extra ?param in the url.

**Comment by imthepitts** [comment]
Hi @eaharonoff,

Sorry for the confusion. When the content type is `application/json`, Twilio injects the `bodySHA256` parameter you're seeing. 

The helper library's `validate()` method will automatically account for this, but it looks like you need to pass the request body as a string in the `data` argument instead of as an array:

https://github.com/twilio/twilio-php/blob/c5fec72e9ce78a15a324e83bc7687a5f89ed29fb/Twilio/Security/RequestValidator.php#L31-L50

Example from our unit test:

https://github.com/twilio/twilio-php/blob/c5fec72e9ce78a15a324e83bc7687a5f89ed29fb/Twilio/Tests/Unit/RequestValidatorTest.php#L43-L49

Please give that a try and let me know if it works for you. We will work on documenting this behavior so that it is clear.

All the Best,
Zack

**Comment by eaharonoff** [comment]
AH you sirs rock @imthepitts @johnpaulmedina. Thanks for all your help. 
One more thing though, and sorry if this is a dumb question but when you say pass the body in as a string, do you mean doing using something like `json_encode($request->all())` or `serialize($request->all())` and pass the result through `validate()`? 
Because I tried this and I found that the first part:
`self::compare( $this->computeSignature($url), $expectedSignature) === true` 
but the latter piece:
`self::compare( $this->computeBodyHash($data), $params['bodySHA256'] ) === false `

**Comment by imthepitts** [comment]
@eaharonoff  can you try accessing the raw request body via PHP instead of using Laravel's middleware? Maybe try `file_get_contents('php://input')`.

**Comment by eaharonoff** [comment]
Really appreciate your help @imthepitts! I've got it working by switching Content Type to `Form URL Encoded` instead 🙌. 
I will say however that I tried to mock your test case exactly but was still getting a mismatch. My hunch is that is coming from the difference in `$this->body()` from your test case, and  `request->all()` in my laravel end-point. `bodySHA256 => 12345` is in my `$request->fullUrl()` as well as the body`$request->all()`. Whereas in your test case, `$this->body()` does not include `bodySHA256`, it is only in `$this->url`.

**Comment by imthepitts** [comment]
@eaharonoff,

I've got this sorted, but looks like a bug in the validator. This portion is expecting the `bodySHA256` to be base64, but it's passed in the URL as hexadecimal and doesn't convert it before comparing:

```
self::compare( $this->computeBodyHash($data), $params['bodySHA256'] ) === false
```

You can get the correct results if you run both comparisons separately and convert `bodySHA256` to base64 before running the second comparison.

Something like this:

```
$hex = $_REQUEST["bodySHA256"];
$base64 = base64_encode(pack('H*', $hex));
$validator->compare( $validator->computeBodyHash($data), $base64 )
```

Will get this in front of the engineering team to resolve in a future release.

All the Best,
Zack

**Comment by childish-sambino** [comment]
Resolved by #523
